src/main/drivers/flash_w25n01g.c

```diff
@@ -580,6 +590,174 @@ void w25n01g_pageProgramFinish(flashDevice_t *fdevice)
         programStartAddress = programLoadAddress;
     }
 }
+#else
+void w25n01g_pageProgramBegin(flashDevice_t *fdevice, uint32_t address, void (*callback)(uint32_t length))
+{
+    fdevice->callback = callback;
+    fdevice->currentWriteAddress = address;
+
+}
+
+static uint32_t currentPage = UINT32_MAX;
+
+// Called in ISR context
+// Check if the status was busy and if so repeat the poll
+busStatus_e w25n01g_callbackReady(uint32_t arg)
+{
+    flashDevice_t *fdevice = (flashDevice_t *)arg;
+    extDevice_t *dev = fdevice->io.handle.dev;
+
+    uint8_t readyPoll = dev->bus->curSegment->u.buffers.rxData[2];
+
+    if (readyPoll & W25N01G_STATUS_FLAG_BUSY) {
+        return BUS_BUSY;
+    }
+
+    // Bus is now known not to be busy
+    fdevice->couldBeBusy = false;
+
+    return BUS_READY;
+}
+
+// Called in ISR context
+// A write enable has just been issued
+busStatus_e w25n01g_callbackWriteEnable(uint32_t arg)
+{
+    flashDevice_t *fdevice = (flashDevice_t *)arg;
+
+    // As a write has just occurred, the device could be busy
+    fdevice->couldBeBusy = true;
+
+    return BUS_READY;
+}
+
+// Called in ISR context
+// Write operation has just completed
+busStatus_e w25n01g_callbackWriteComplete(uint32_t arg)
+{
+    flashDevice_t *fdevice = (flashDevice_t *)arg;
+
+    fdevice->currentWriteAddress += fdevice->callbackArg;
+
+    // Call transfer completion callback
+    if (fdevice->callback) {
+        fdevice->callback(fdevice->callbackArg);
```

Comment:
We only update `bufferTail` in `flashfsAdvanceTailInBuffer` (except following an erase) so I think this is safe. Can't think of a failure case, but perhaps I'm not thinking hard enough.

Reply:
Problem is with reading it (flashfsGetDirtyDataBuffers is affected too).
```

bufferHead = 5;
bufferTail = 9;
FLASHFS_WRITE_BUFFER_SIZE  = 10;

static uint32_t flashfsTransmitBufferUsed(void)
{
    if (bufferHead >= bufferTail)
        return bufferHead - bufferTail;
    // ISR update bufferTail, incrementing it by 1 bytes and wrapping around, bufferTail == 0
    return FLASHFS_WRITE_BUFFER_SIZE - bufferTail + bufferHead;
    // 15 is returned 
}
```

In this case, fix is trivial. Function may return slightly old value, but that is expected:
```
static uint32_t flashfsTransmitBufferUsed(void)
{
    unsigned tail = bufferTail;
    if (bufferHead >= tail)
        return bufferHead - tail;

    return FLASHFS_WRITE_BUFFER_SIZE - tail + bufferHead;
}
```

The problem is not that obvious, because `flashfsWriteByte` is even more buggy and will silently wraparound buffer. And blackbox tries to prevent overflowing buffers, so this is not hit too often.

Reply:
(I have learned to take these problems very seriously - bug happening once in a century (per device) is much worse than the one happening every day, ale least if you manage thousands of  devices)